### PR TITLE
[innernet] wrapper for wireguard

### PR DESCRIPTION
**Pull request by plan5**
[innernet](https://github.com/tonarino/innernet) helps setting up and with administration of wireguard networks.

This package currently puts config in /etc/innernet. There is a commented line in the prepare section to change this to /opt/etc/innernet.

For now, I create this PR so others can test it. innernet is available as a community package in Arch or as deb packages on the repo releases page.

To Do:

- [x] Add (installation of) a service file
- [x] See if wireguard package will move config to /opt/etc and change innernet recipe accordingly

### CODE REVIEW

**Comment by Eeems** [comment]
https://github.com/tonarino/innernet/blob/6d28e7f4ab20800e30ca759727e9cfbd9d6da17c/shared/src/lib.rs#L19-L22
I'd suggest that all of these should be moved under `/opt` to play nice with OS upgrades.

**Comment by plan5** [comment]
> https://github.com/tonarino/innernet/blob/6d28e7f4ab20800e30ca759727e9cfbd9d6da17c/shared/src/lib.rs#L19-L22
> I'd suggest that all of these should be moved under `/opt` to play nice with OS upgrades.

Done. Does the wireguard package keep its configs in /etc though?

**Comment by Eeems** [comment]
> Done. Does the wireguard package keep its configs in /etc though?

Not if you modify that.

**Comment by plan5** [comment]
> > Done. Does the wireguard package keep its configs in /etc though?
> 
> Not if you modify that.

Ok, I've changed it for the innernet configs but not for wireguard because that would have to be done in the wireguard package first.

I'll also have to add installation of a service file

**Comment by plan5** [comment]
@Eeems Since the wireguard package still has its configs in /etc/, I'll only change the location of the innernet files for now.

Should I write an issue for wireguard to change its config location before we move on with this package?

**Comment by Eeems** [comment]
> Ok, I've changed it for the innernet configs but not for wireguard because that would have to be done in the wireguard package first.
> 
> I'll also have to add installation of a service file

> @Eeems Since the wireguard package still has its configs in /etc/, I'll only change the location of the innernet files for now.
> 
> Should I write an issue for wireguard to change its config location before we move on with this package?



Sorry I was dealing with a headache while I wrote that reply. Didn't even realize I had sent it. Wireguard is still likely using `/etc` which is something we probably should patch in a different PR.

File: package/innernet/package [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    sed -i "s/\/etc\/innernet/\/opt\/etc\/innernet/g"  "$srcdir/shared/src/lib.rs"
+
+    #tbd: change config dir to /opt/etc/innernet?
+    sed -i "s/\/var\/lib\/innernet/\/opt\/etc\/innernet/g"  "$srcdir/shared/src/lib.rs"
+
+    #Change binary path in service file
+    sed -i "s/\/usr\/bin\/innernet/\/opt\/bin\/innernet/g"  "$srcdir/client/innernet@.service"
+
+    #tbd: change wireguard config dir to /opt/etc/wireguard?
+    #sed -i "s/\/etc\/wireguard/\/opt\/etc\/wireguard/g"  "$srcdir/shared/src/lib.rs"
```
Comment (Eeems) [inline package/innernet/package]:
```suggestion
    # Change config folders to live under /opt
    sed -i "s/\/etc\/innernet/\/opt\/etc\/innernet/g" "$srcdir/shared/src/lib.rs"
    sed -i "s/\/var\/lib\/innernet/\/opt\/etc\/innernet/g" "$srcdir/shared/src/lib.rs"

    #Change binary path in service file
    sed -i "s/\/usr\/bin\/innernet/\/opt\/bin\/innernet/g" "$srcdir/client/innernet@.service"

    #tbd: change wireguard config dir to /opt/etc/wireguard?
    #sed -i "s/\/etc\/wireguard/\/opt\/etc\/wireguard/g" "$srcdir/shared/src/lib.rs"
```

**Comment by plan5** [comment]
> > Ok, I've changed it for the innernet configs but not for wireguard because that would have to be done in the wireguard package first.
> > I'll also have to add installation of a service file
> 
> > @Eeems Since the wireguard package still has its configs in /etc/, I'll only change the location of the innernet files for now.
> > Should I write an issue for wireguard to change its config location before we move on with this package?
> 
> Sorry I was dealing with a headache while I wrote that reply. Didn't even realize I had sent it. Wireguard is still likely using `/etc` which is something we probably should patch in a different PR.

No problem :) I was just a bit confused. Get well!

**Comment by Eeems** [comment]
Testing on the rM1 and it looks like you haven't put in the shorthand for `innernet` (`inn`). Could you add that symlink as well?

Other than that it works!
image

**Comment by plan5** [comment]
> Testing on the rM1 and it looks like you haven't put in the shorthand for `innernet` (`inn`). Could you add that symlink as well?
> 
> Other than that it works!

Done! I've added the link.

Also nice to hear it works for you! :) I'm having problems pinging between peers but I guess that's a server misconfig then. Befor your post I didn't know if it's because of the package.

Would be interesting to know if it works for someone else with rM2 though...

**Comment by Eeems** [comment]
> Also nice to hear it works for you! :) I'm having problems pinging between peers but I guess that's a server misconfig then. Befor your post I didn't know if it's because of the package.
> 
> Would be interesting to know if it works for someone else with rM2 though...

@danshick or @raisjn would you be able to test? You'll need to go through setting up a innernet server to test with as well. If you need help with that feel free to ping me, or ask on the innernet discord: https://discord.gg/fs2nqxWKhx


> Done! I've added the link.

Awesome, I'll test that sometime soon.

**Comment by danshick** [comment]
I should be able to give it a shot along with your oxide PRs tonight.

Edit: Got delayed yesterday and I only had about a half hour today. Planning to take a look tomorrow instead.
Edit2: And one day became four

**Comment by mcginty** [comment]
Author of innernet here! I'm happy to add a CLI option in the next release to override the default config and data directories so you don't have to maintain patches.

**Comment by danshick** [comment]
> Author of innernet here! I'm happy to add a CLI option in the next release to override the default config and data directories so you don't have to maintain patches.

That'd be great! Thanks for your work.

**Comment by Eeems** [comment]
Alright, so I can confirm that the link for `inn` works, but I've encountered an issue. Since I've rebooted my device, innernet is failing to start
image

@mcginty any ideas how I can get more information out of this error?

**Comment by mcginty** [comment]
@Eeems hmm, I need to add better error messages! If you run `inn up eeems` manually, that has the same issue I assume? Is the wireguard kernel module loaded properly (`lsmod | grep wireguard`)?

**Comment by Eeems** [comment]
Correct it does give the same error message. I forgot to check the kernel module to make sure it was loaded. Likely I forgot to add it to anything to handle it automatically loading on startup. I'll check again tomorrow and let you know.

**Comment by Eeems** [comment]
@mcginty Okay, tried again and it is still failing after confirming that wireguard is loaded: 
image

**Comment by plan5** [comment]
It's working over reboots for me on rM2. 
Just adding for documentation's sake. 
Could this be related to rM1 exclusively?

**Comment by Eeems** [comment]
> It's working over reboots for me on rM2.
> Just adding for documentation's sake.
> Could this be related to rM1 exclusively?

Potentially, since the rM1 loads wireguard from a different location, see https://github.com/toltec-dev/toltec/blob/stable/package/wireguard/package#L26-L29

**Comment by mcginty** [comment]
I added more logging in [v1.3.0-beta.3](https://github.com/tonarino/innernet/releases/tag/v1.3.0-beta.3) that should hopefully help with debugging the issue on this platform. Maybe give that a shot?

**Comment by plan5** [comment]
should we maybe push this to beta.4?

@mcginty is the additional loggin only available in beta.3 or also in beta.4?
I guess that decides it.

**Comment by Eeems** [comment]
image
Alright this is interesting, how did `/opt/var/lib/innernet` get removed? Upon creating the directory it works fine.
image
So for now the package should ensure the folder exists on configure. I'll open an issue with innernet itself to get it created.

File: package/innernet/package [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+package() {
+    install -D -m 700 -t "$pkgdir"/opt/bin "$srcdir"/target/armv7-unknown-linux-gnueabihf/release/innernet
+    install -D -m 644 -t "$pkgdir"/lib/systemd/system "$srcdir/client/innernet@.service"
+    install -D -m 644 -t "$pkgdir"/lib/systemd/system "$srcdir/client/innernet.target"
+
+    # Link /opt/bin/innernet to /opt/bin/inn
+    ln -s /opt/bin/innernet "$pkgdir"/opt/bin/inn
+}
```
Comment (Eeems) [inline package/innernet/package:43]:
```suggestion
}

configure() {
    if ! [ -d /opt/var/lib/innernet ]; then
        mkdir -p /opt/var/lib/innernet
    fi
}
```

**Review by Eeems — CHANGES_REQUESTED** [review]
Ensure that the data directory (`/opt/lib/innernet`) exists on install/upgrade.

**Review by Eeems — APPROVED** [review]

**Pull request merged**